File: Realize/9Week/PracticeSort1.cpp
#include <iostream>
#include <vector>
using namespace std;

void printArray(const vector<int>& arr, const string& name) {
    cout << name << ": ";
    for (int n : arr) {
        cout << n << " ";
    }
    cout << endl;
}

// 버블 정렬
void bubbleSort(vector<int>& arr) {
    for (int i = 1; i < arr.size() - 1; i++) {
        for (int j = 0; j < arr.size() - i; j++) {
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
            }
        }
    }

    for (int i = arr.size() - 1; i > 0; i--) {
        for (int j = 0; j < i; j++) {
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
            }
        }
    }
}

// 삽입 정렬
void insertionSort(vector<int>& arr) {
    for (int i = 1; i < arr.size(); i++) {
        for (int j = i; j > 0; j--) {
            if (arr[j] < arr[j - 1]) {
                swap(arr[j], arr[j - 1]);
            } else {
                break;
            }
        }
    }
}

// 선택 정렬
void selectionSort(vector<int>& arr) {
    for (int i = 0; i < arr.size() - 1; i++) {
        int minIdx = i;
        for (int j = i + 1; j < arr.size(); j++) {
            if (arr[j] < arr[minIdx]) {
                minIdx = j;
            }
        }
        swap(arr[i], arr[minIdx]);
    }

    for (int i = arr.size() - 1; i > 0; i--) {
        int maxIdx = i;
        for (int j = i - 1; j >= 0; j--) {
            if (arr[j] > arr[maxIdx]) {
                maxIdx = j;
            }
        }
        swap(arr[i], arr[maxIdx]);
    }
}

int main() {
    vector<int> arr = {3, 5, 2, 7, 1, 4};

    vector<int> temp = arr;
    bubbleSort(temp);
    printArray(temp, "버블 정렬");

    temp = arr;
    insertionSort(temp);
    printArray(temp, "삽입 정렬");

    temp = arr;
    selectionSort(temp);
    printArray(temp, "선택 정렬");

    return 0;
}

File: Realize/9Week/PracticeSort1.py
def bubble_sort(arr):
     # 1.
    for i in range(1, len(arr) - 1):
        for j in range(len(arr) - i):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
    # 2. (1과 2 같은 방식인데 i 인덱스 사용 다르게 참고)
    for i in range(len(arr) - 1, 0, -1):
        for j in range(i):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]

def insertion_sort(arr):
    for i in range(1, len(arr)):
        for j in range(i, 0, -1):
            if arr[j] < arr[j - 1]:
                arr[j], arr[j - 1] = arr[j - 1], arr[j]
            else:
                break

def selection_sort(arr):
     # 1. 최소 값을 찾아 앞 쪽부터 교환하는 방식
    for i in range(len(arr) - 1):
        min_idx = i
        for j in range(i + 1, len(arr)):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
     # 2. 최대 값을 찾아 뒤 쪽부터 교환하는 방식
    for i in range(len(arr) - 1, 0, -1):
        max_idx = i
        for j in range(i - 1, -1, -1):
            if arr[j] > arr[max_idx]:
                max_idx = j
        arr[i], arr[max_idx] = arr[max_idx], arr[i]

# Test code
arr = [3, 5, 2, 7, 1, 4]

temp = arr[:]
bubble_sort(temp)
print("버블 정렬:", temp)

temp = arr[:]
insertion_sort(temp)
print("삽입 정렬:", temp)

temp = arr[:]
selection_sort(temp)
print("선택 정렬:", temp)
